src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,441 @@
-      get() match {
-        case linked: DefaultPromise[_] => compressedRoot(linked)
-        case _ => this
-      }
+    override final def filter(@deprecatedName('pred) p: T => Boolean)(implicit executor: ExecutionContext): Future[T] = {
+      val state = get()
+      if (!state.isInstanceOf[Failure[T]]) dispatchOrAddCallbacks(state, new Transformation[T, T](Xform_filter, p, executor)) // Short-circuit if we get a Success
+      else this
+    }
 
-    @tailrec
-    private[this] final def compressedRoot(linked: DefaultPromise[_]): DefaultPromise[T] = {
-      val target = linked.asInstanceOf[DefaultPromise[T]].root
-      if (linked eq target) target
-      else if (compareAndSet(linked, target)) target
-      else {
-        get() match {
-          case newLinked: DefaultPromise[_] => compressedRoot(newLinked)
-          case _ => this
-        }
-      }
+    override final def collect[S](pf: PartialFunction[T, S])(implicit executor: ExecutionContext): Future[S] = {
+      val state = get()
+      if (!state.isInstanceOf[Failure[T]]) dispatchOrAddCallbacks(state, new Transformation[T, S](Xform_collect, pf, executor)) // Short-circuit if we get a Success
+      else this.asInstanceOf[Future[S]]
     }
 
-    /** Get the promise at the root of the chain of linked promises. Used by `compressedRoot()`.
-     *  The `compressedRoot()` method should be called instead of this method, as it is important
-     *  to compress the link chain whenever possible.
-     */
-    @tailrec
-    private def root: DefaultPromise[T] =
-      get() match {
-        case linked: DefaultPromise[_] => linked.asInstanceOf[DefaultPromise[T]].root
-        case _ => this
-      }
+    override final def recoverWith[U >: T](pf: PartialFunction[Throwable, Future[U]])(implicit executor: ExecutionContext): Future[U] = {
+      val state = get()
+      if (!state.isInstanceOf[Success[T]]) dispatchOrAddCallbacks(state, new Transformation[T, U](Xform_recoverWith, pf, executor)) // Short-circuit if we get a Failure
+      else this.asInstanceOf[Future[U]]
+    }
+
+    override final def recover[U >: T](pf: PartialFunction[Throwable, U])(implicit executor: ExecutionContext): Future[U] = {
+      val state = get()
+      if (!state.isInstanceOf[Success[T]]) dispatchOrAddCallbacks(state, new Transformation[T, U](Xform_recover, pf, executor)) // Short-circuit if we get a Failure
+      else this.asInstanceOf[Future[U]]
+    }
+
+    override final def mapTo[S](implicit tag: scala.reflect.ClassTag[S]): Future[S] =
+      if (!get.isInstanceOf[Failure[T]]) super[Future].mapTo[S](tag) // Short-circuit if we get a Success
+      else this.asInstanceOf[Future[S]]
+
+    override final def failed: Future[Throwable] = {
+      val state = get()
+      if (!state.isInstanceOf[Success[T]]) super.failed
+        //dispatchOrAddCallbacks(state, new Transformation[T, Throwable](Xform_failed, Future.failedFun[T], InternalCallbackExecutor)) // Short-circuit if we get a Failure
```

Comment:
What's the status of this commented-out line?

Reply:
@NthPortal Nice catch—it's a leftover from me experimenting what operations made sense to overload.